new-frontiers-14/frontier-station-14
# Handheld Radio Frequencies

## About the PR



Based off of [nuclear-14#390](https://github.com/Vault-Overseers/nuclear-14/pull/390/files) with a modest set of changes so far.  Thank you to both poeMota for the original feature, and Sh1ntra for porting the feature to Corvax & nuclear-14.

Adds selectable frequencies to handheld radios, and a control interface similar to the intercom's.

Selectable frequencies are currently 1000-3000 ([assuming values are in kHz, radio buffs weep](https://en.wikipedia.org/wiki/Radio_spectrum)).  Invalid values being entered will result in the frequency being queried.

TODOs:

- [X] Add frequency to ghost/admin chat logs for handheld radios.
- [ ] ~~Better select which radio you're using than whatever enumeration currently exists.~~ Larger rework needed here, probably not worth it, at least for now.
- [X] Restrict frequencies to some range that makes sense (500 - 3000?  Reasonable values in kHz, allows for 2559 for the memes.)
- [X] Prevent admins being spammed by handheld radio output.

## Why / Balance


Useful for different crews, prevents crosstalk on the same radio channel.

## How to test


1. Join on one client (Client 1), observe as admin in another (Client 2).
2. On Client 1, spawn a few handheld radios, pick up one.  Turn on the speakers on for all of them, and turn on the mic for the one in your hand.
3. On Client 1, say a message, you should hear it on all nearby radios.  On Client 2, you should hear the message said by the user in local chat, you should hear the message on the Handheld channel on the proper frequency, and you should hear the message whispered by each radio _that is close to your ghost_.
4. On Client 1, change the frequency for the radio in your hand.  Say a message, you should not hear it back from any other radio.
5. On Client 1, change the frequency for one of the other radios to the frequency you set in step 4.  Say a message, you should hear it back from that radio.
6. On Client 1, change the frequency for _another_ of the radios to a separate frequency and turn on its microphone.  Say a message, it should be transmitted _by one of the radios_ - this order seems unpredictable, but it is first come, first serve for the event handler.
7. On Client 1, try to set the frequency to invalid values.  Any values too low, too high, non-numeric values should be handled appropriately.  Any invalid value should not result in a change to the last selected frequency.

## Media


Radio UI: elements have a better fit, flavour text shows usable frequencies.
![image](https://github.com/user-attachments/assets/62b4cb8d-8433-498d-8c8a-2e794fdde974)

Messages received as a ghost.  The whispers are only visible when near the source of the message.
![image](https://github.com/user-attachments/assets/db3d4cb1-614a-4b74-9a44-61d2c60053dc)


## Breaking changes


**Changelog**



:cl: poeMota, Sh1ntra, whatston3
- add: Handheld radios have been replaced with HandiComms, now with selectable radio frequencies.
